### PR TITLE
Add King's Rock flags for Gens 2-4

### PR DESCRIPTION
**Pull request by bgsamm**
As discussed in issue #3343, Gens 2-4 use "move scripts" to allow King's Rock to proc for specific moves. This pull adds mods to said gens to reflect this. Sources used are [UPC](https://web.archive.org/web/20170530082915/http://upcarchive.playker.info:80/0/upokecenter/content/pokemon-gold-version-silver-version-and-crystal-version-list-of-moves.html) (linked in the relevant task in #2367) and Bulbapedia (for Gen 4, since there is no archive for the Gen 4 move list on UPC). Additionally, this pull adds checks to Gen 2 to limit multi-hit moves to a single roll for status and to prevent moves from flinching sleeping/frozen targets unless the user is holding a King's Rock.

Notes:
* The multi-hit check, which is mainly for King's Rock, also limits Twineedle to a single roll for poison, which I believe is accurate
* Snore is affected by King's Rock, but already has a chance to flinch. I was unsure how this scenario is handled in-game, both in flinch chance and in ability to flinch sleeping/frozen targets. (There are a few other such moves in Gens 3 & 4, such as Rolling Kick, Twister, and Dark Pulse)

### CODE REVIEW

**Comment by Zarel** [comment]
Do you know when the single roll takes place? I'd assume after all hits, right? So the "poisoned" message would only happen after Twineedle's last hit?

**Comment by bgsamm** [comment]
> Do you know when the single roll takes place? I'd assume after all hits, right? So the "poisoned" message would only happen after Twineedle's last hit?

Mm, yea, good point. I'll try to fix that.

**Comment by Marty-D** [comment]
For Gen 4, I checked UPC against your list from Bulbapedia and found these changes to be made:
```diff
+ Dragon Rage, Gust, Pound, Wring Out, Zen Headbutt
- Tickle
```

I know Tickle is actually flagged in Gen 3 for whatever reason, and for Gen 4 Zen Headbutt sounded odd so I went and checked a disassembly. https://github.com/JimB16/PokePlat/blob/master/data/poketool/waza/waza_tbl/data_00000428.s
Zen Headbutt is index number 428
Flag2 = 51 = 0b00110011 = UPC notation `||fe||ba` = kingsrock/mirror/protect/contact


Gen 4 also introduced Razor Fang, so whatever is done with King's Rock should be copied there too.

> Do you know when the single roll takes place? I'd assume after all hits, right? So the "poisoned" message would only happen after Twineedle's last hit?

While this sounds right, if Gen 2 Twineedle's first hit breaks a substitute, the second hit can't poison. PS already handles this by deleting its secondaries when it hits a sub, but UPC also says flinches from King's Rock can happen if a sub is broken, so that may need to be revisited. The Crystal disassembly doesn't seem to back up that claim, but I'm by no means an expert on reading asm. I'll try to test that in-game soon.


Re: King's Rock and moves that can already flinch, for Gens 3 and 4 UPC says "(This effect is cumulative with additional effects that already cause their target to flinch.)", which to me reads as being a completely independent roll since effect chances have never been additive.

For Gen 2 and 3 I'll double check UPC's flags with the pret disassemblies later. Overall this is looking really good! :+1:

**Comment by bgsamm** [comment]
Cool. I've updated the Gen 4 flags, and I'll change the multi-hit check to after the last hit instead of the first. I've also copied the flags over to Razor Fang for Gen 4, under the assumption that the list is the same as King's Rock. I guess I'll also need to implement a `kingsrock` flag or something to the secondary, to deal with e.g. Snore. I'll wait to commit though until you double check the Gen 2 and 3 flags. I'll also leave the sub interaction alone for the time-being.

One other thing, I was confused by the original Gen 3 King's Rock mod; was there a specific reason that it wasn't adding the King's Rock secondary if the move already had secondaries? Was that just an early attempt at limiting it to the right moves?

**Comment by Marty-D** [comment]
Well, this sucks:
kingsrock
Feel free not to deal with that in this PR since it'll involve refactoring some other stuff.

Your lists for Gen 2 and 3 moves match the respective game's code, so all's good there.

>Was that just an early attempt at limiting it to the right moves?

As far as I know, yeah. The person who implemented Gen 3 several years ago used a combination of UPC and Bulbapedia, so some things were originally done wrong or half-wrong.

**Comment by Marty-D** [comment]
I rounded the 11.7% chance in Gen 2 up to 12% so that when it's translated into "chance out of 256", it'll round down to 30 instead of 29.

This looks well done to me! @Zarel did you have any other comments?

File: mods/gen2/moves.js [hunk 1]
```diff
@@ -842,7 +842,10 @@ let BattleMovedex = {
 					this.debug('sub bypass: self hit');
 					return;
 				}
-				if (move.id === 'twineedle') delete move.secondaries;
```
Comment (Zarel) [inline mods/gen2/moves.js]: Hm, maybe this should just always be defined, at least for an ActiveMove.

**Comment by Zarel** [comment]
This looks good!

**Pull request merged**